### PR TITLE
lxd: always (re-)injects snaps

### PR DESCRIPTION
**Pull request by kalikiana**
- [x] Have you followed the [guidelines for contributing](https://github.com/snapcore/snapcraft/blob/master/CONTRIBUTING.md)?
- [x] Have you signed the [CLA](http://www.ubuntu.com/legal/contributors/)?
- [x] If this is a bugfix. Have you checked that there is a bug report open for the issue you are trying to fix on [bug reports](https://bugs.launchpad.net/snapcraft)?
- [x] If this is a new feature. Have you discussed the design on the [forum](https://forum.snapcraft.io)?
- [x] Have you successfully run `./runtests.sh static`?
- [x] Have you successfully run `./runtests.sh unit`?

-----
Fixes: [LP: #1741752](https://bugs.launchpad.net/snapcraft/+bug/1741752)

### CODE REVIEW

File: snapcraft/tests/unit/test_lxd.py [hunk 1]
```diff
@@ -537,6 +537,49 @@ def make_containerbuild(self):
...
+            {'name': 'snapcraft',
+             'confinement': 'classic',
+             'id': '3lljuRvshPlM4gpJnH5xExo0DJBOvImu',
+             'channel': 'edge',
+             'revision': '345'},
+        ]
+
+        self.make_containerbuild().execute()
+
+        if hasattr(self, 'cross') and self.cross:
```
Comment (kyrofa) [inline snapcraft/tests/unit/test_lxd.py]: What is this doing, exactly? What is this attribute?

File: snapcraft/tests/unit/test_lxd.py [hunk 2]
```diff
@@ -537,6 +537,49 @@ def make_containerbuild(self):
                            project_options=self.project_options,
                            remote=self.remote)
 
+    @patch('snapcraft.internal.lxd.Containerbuild._container_run')
+    @patch('snapcraft.internal.common.is_snap')
+    def test_inject_snap_existing_container(self,
```
Comment (kyrofa) [inline snapcraft/tests/unit/test_lxd.py]: What makes this container already-existing?

File: snapcraft/internal/lxd/_containerbuild.py [hunk 3]
```diff
@@ -226,13 +226,13 @@ def _wait_for_network(self):
...
 
-    def _inject_snapcraft(self):
+    def _inject_snapcraft(self, *, new_container: bool):
         if common.is_snap():
             with tempfile.TemporaryDirectory(
                     prefix='snapcraft', dir=self._lxd_common_dir) as tmp_dir:
                 self._inject_snap('core', tmp_dir)
                 self._inject_snap('snapcraft', tmp_dir)
-        else:
+        elif new_container:
```
Comment (kyrofa) [inline snapcraft/internal/lxd/_containerbuild.py]: I feel like I'm missing something, here. How does this re-inject the snap?

File: snapcraft/tests/unit/test_lxd.py (hunk 2 above)
Comment (kalikiana) [reply snapcraft/tests/unit/test_lxd.py]: Apparently I didn't commit that bit... fixed

File: snapcraft/tests/unit/test_lxd.py (hunk 1 above)
Comment (kalikiana) [reply snapcraft/tests/unit/test_lxd.py]: This is using the same approach as the other test cases in this file. The snaps are installed from the store and not from a local file in the case of a remote with another architecture.

Comment (kalikiana) [reply snapcraft/tests/unit/test_lxd.py]: Oh, I didn't realize they're not used in the ProjectTestCase. Adding.

**Comment by kalikiana** [comment]
@sergiusens Note: this branch is about injecting snaps into existing containers. The `_inject_snap` code is the same except for replacing the `_container_arch` variable with a method. So tweaks of that logic should potentially go into a separate branch.
The file movements come first because 1) x-revisioned snaps aren't user-accessible 2) we compare the sha384sum since we can't universally rely on revisions. We would need to have a separate code path for store-owned snaps (would have to check if snapd talks HTTP, or use `snap info`).

I don't understand your point about confusing results. Assertions are not relevant to side-loaded snaps, and only pushed before actual installation of the snap.

File: snapcraft/internal/lxd/_containerbuild.py [hunk 4]
```diff
@@ -298,6 +289,14 @@ def _inject_snap(self, name: str, tmp_dir: str):
                            is_dangerous=rev.startswith('x'),
                            is_classic=is_classic)
 
+    def _get_container_arch(self):
```
Comment (kyrofa) [inline snapcraft/internal/lxd/_containerbuild.py]: Good call, moving this to be checked every time.

File: snapcraft/tests/unit/test_lxd.py [hunk 5]
```diff
@@ -537,6 +537,52 @@ def make_containerbuild(self):
...
+                                            mock_is_snap,
+                                            mock_container_run):
+        mock_is_snap.side_effect = lambda: True
```
Comment (kyrofa) [inline snapcraft/tests/unit/test_lxd.py]: Why not just `mock_is_snap.return_value = True`?

File: snapcraft/tests/unit/test_lxd.py [hunk 6]
```diff
@@ -537,6 +537,52 @@ def make_containerbuild(self):
...
+        mock_container_run.side_effect = lambda cmd, **kwargs: cmd
```
Comment (kyrofa) [inline snapcraft/tests/unit/test_lxd.py]: The real function doesn't actually return anything... is this really what you want?

File: snapcraft/tests/unit/test_lxd.py [hunk 7]
```diff
@@ -537,6 +537,52 @@ def make_containerbuild(self):
...
+        if hasattr(self, 'cross') and self.cross:
+            mock_container_run.assert_has_calls([
+                call(['snap', 'install', 'core', '--channel', 'stable']),
+                call(['snap', 'refresh', 'core', '--channel', 'stable']),
+                call(['snap', 'install', 'snapcraft', '--channel', 'edge',
+                      '--classic']),
+                call(['snap', 'refresh', 'snapcraft', '--channel', 'edge',
+                      '--classic']),
+            ])
+            return
```
Comment (kyrofa) [inline snapcraft/tests/unit/test_lxd.py]: I'd rather see an `else` here instead of an early exit.

**Pull request merged**